openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -1056,6 +1072,150 @@ def revert_changes(library_key):
+        library_collection=LibraryCollectionData(
+            library_key=library_key,
+            collection_id=collection.id
+        )
+    )
+
+    return collection
+
+
+def update_library_collection_components(
+    library_key: LibraryLocatorV2,
+    collection_id: int,
+    usage_keys: list[UsageKeyV2],
+    created_by: int | None = None,
+    remove=False,
+    content_library: ContentLibrary | None = None,
+) -> Collection:
+    """
+    Associates the Collection with Components for the given UsageKeys.
+
+    By default the Components are added to the Collection.
+    If remove=True, the Components are removed from the Collection.
+
+    If you've already fetched the ContentLibrary, pass it in to avoid refetching.
+
+    Raises:
+    * ContentLibraryCollectionNotFound if no Collection with the given pk is found in the given library.
+    * ContentLibraryBlockNotFound if any of the given usage_keys don't match Components in the given library.
+
+    Returns the updated Collection.
+    """
+    if not content_library:
+        content_library = ContentLibrary.objects.get_by_key(library_key)  # type: ignore[attr-defined]
+    assert content_library
+    assert content_library.learning_package_id
+    assert content_library.library_key == library_key
+
+    # Fetch the Component.key values for the provided UsageKeys.
+    component_keys = []
+    for usage_key in usage_keys:
+        # Parse the block_family from the key to use as namespace.
+        block_type = BlockTypeKey.from_string(str(usage_key))
+
+        try:
+            component = authoring_api.get_component_by_key(
+                content_library.learning_package_id,
+                namespace=block_type.block_family,
+                type_name=usage_key.block_type,
+                local_key=usage_key.block_id,
+            )
+        except Component.DoesNotExist as exc:
+            raise ContentLibraryBlockNotFound(usage_key) from exc
+
+        component_keys.append(component.key)
+
+    # Note: Component.key matches its PublishableEntity.key
+    entities_qset = PublishableEntity.objects.filter(
+        key__in=component_keys,
+    )
+
+    if remove:
+        collection = authoring_api.remove_from_collection(
+            collection_id,
+            entities_qset,
+        )
+    else:
+        collection = authoring_api.add_to_collection(
+            collection_id,
+            entities_qset,
+            created_by=created_by,
+        )
+
+    # Emit event for library collection updated
+    LIBRARY_COLLECTION_UPDATED.send_event(
+        library_collection=LibraryCollectionData(
+            library_key=library_key,
+            collection_id=collection_id
+        )
+    )
+
+    # Emit a CONTENT_OBJECT_TAGS_CHANGED event for each of the objects added/removed
+    for usage_key in usage_keys:
+        CONTENT_OBJECT_TAGS_CHANGED.send_event(
+            content_object=ContentObjectData(object_id=usage_key),
+        )
```

Comment:
@bradenmacdonald Since the `collection.modified` date changes here, I'm emitting a `LIBRARY_COLLECTION_UPDATED` event so the collection's metadata can be reindexed.

@rpenido [suggested on openedx-events#386](https://github.com/openedx/openedx-events/pull/386#issuecomment-2316518199) that, instead of emitting all of these `CONTENT_OBJECT_TAGS_CHANGED` events, we could let the `LIBRARY_COLLECTION_UPDATED` handler reindex each of the collection's content objects. I could gate this using your suggested `reasons` field (or with just a boolean flag?) to see whether or not the entities have changed.

Then we could leave `CONTENT_OBJECT_TAGS_CHANGED` as-is.

What would you prefer?

Reply:
I'd prefer to emit a `CONTENT_OBJECT_TAGS_CHANGED` (or `CONTENT_OBJECT_ASSOCIATIONS_CHANGED`) for each individual component. It's much more efficient to just push an updated document for that one component into the search index, rather than to re-index the entire collection. We want to avoid re-indexing a whole collection/library/course as much as possible. (I think the only time we currently _have_ to do it is on publishing or discarding changes, because we don't have a good way to know which individual components to reindex in that case.)

It's also totally fine to emit many small events. They'll get enqueued and processed in order, and it doesn't matter if it's slightly less efficient in some cases than one "big" event.